### PR TITLE
Create base classes for block and variable events

### PR DESCRIPTION
**Pull request by rachel-fenichel**
...instead of stuffing everything into Abstract

## The basics

- [x] I branched from develop
- [x] My pull request is against develop
- [x] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)

### Resolves

None

### Proposed Changes

Create base classes for variable events and block events.  Remove things from Events.Abstract that depend on checking the type.

### Reason for Changes

As I added another type of events (comments) I realized that we had lots of fields in Events.Abstract that are only conditionally set, depending on the type of the event.  That means that adding new event types involves moreinstanceof/typeof checks in Abstract, which I'm not a fan of.

Instead I have a base class for each event type.  Fields that are required to be set in that type of event are set in the base class: blockId for blocks, varId for variables, etc.

### Test Coverage
Jsunit tests pass.

### CODE REVIEW

File: core/events/abstract.js [hunk 1]
```diff
@@ -33,24 +33,9 @@ goog.require('goog.math.Coordinate');
 
 /**
  * Abstract class for an event.
- * @param {Blockly.Block|Blockly.VariableModel} elem The block or variable.
  * @constructor
  */
-Blockly.Events.Abstract = function(elem) {
```
[the remark this answers is not in the document]
Comment (rachel-fenichel) [reply core/events/abstract.js]: Yeah, blockId is already public and has been for a while.  varId has been public for about a year.

File: core/events/abstract.js [hunk 2]
```diff
@@ -63,22 +48,12 @@ Blockly.Events.Abstract = function(elem) {
...
 
   /**
    * Sets whether the event should be added to the undo stack.
    * @type {boolean}
    */
-  this.recordUndo = undefined;
-
-  if (elem instanceof Blockly.Block) {
-    this.blockId = elem.id;
-    this.workspaceId = elem.workspace.id;
```
[the remark this answers is not in the document]
Comment (rachel-fenichel) [reply core/events/abstract.js]: It's always set but comes from somewhere different in each case.

File: core/events/ui/ui_base.js [hunk 3]
```diff
@@ -41,7 +41,13 @@ goog.require('goog.math.Coordinate');
  * @constructor
  */
 Blockly.Events.Ui = function(block, element, oldValue, newValue) {
-  Blockly.Events.Ui.superClass_.constructor.call(this, block);
+  if (!block) {
+    return;  // Blank event to be populated by fromJson.
```
[the remark this answers is not in the document]
Comment (rachel-fenichel) [reply core/events/ui/ui_base.js]: To create an event from JSON we create an event of that type, then call `.fromJSON(json)` on it, passing in the JSON to parse.
It does kind of feel like it should be a static function: `Blockly.Events.EventType.fromJSON(json)` that returns a newly constructed `Blockly.Events.EventType` object.

**Comment by rachel-fenichel** [comment]
Tested under advanced compilation and it works.  Good call.

File: core/events/ui/ui_base.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (picklesrus) [reply core/events/ui/ui_base.js]: but, what does that have to do with the block being null/undefined?

[the remark this answers is not in the document]
Comment (rachel-fenichel) [reply core/events/ui/ui_base.js]: See `Blockly.Events.fromJson` (https://github.com/google/blockly/blob/develop/core/events/events.js#L303)

The way to do this is currently
```
event = new Blockly.Events.MyType(null);
event.fromJson(json);
```

**Pull request merged**